prow/cmd/generic-autobumper/bumper/bumper.go

```diff
@@ -891,3 +937,139 @@ func getTreeRef(stderr io.Writer, refname string) (string, error) {
+	if err := Call(stdout, stderr, gitCmd, "config", "http.cookiefile", cookieFile); err != nil {
+		return fmt.Errorf("unable to load cookiefile: %v", err)
+	}
+	if err := Call(stdout, stderr, gitCmd, "remote", "add", "upstream", hostRepo); err != nil {
+		return fmt.Errorf("unable to add upstream remote: %v", err)
+	}
+	noOp, err := GerritNoOpChange(changeID, hostRepo)
+	if err != nil {
+		return fmt.Errorf("error diffing previous bump: %v", err)
+	}
+	if noOp {
+		logrus.Info("CR is a no-op change. Returning without pushing update")
+		return nil
+	}
+
+	pushRef := buildPushRef(branch, reviewers, cc)
+	if err := Call(stdout, stderr, gitCmd, "commit", "-a", "-v", "-m", msg); err != nil {
+		return fmt.Errorf("unable to commit: %v", err)
+	}
+	if err := Call(stdout, stderr, gitCmd, "push", "origin", pushRef); err != nil {
+		return fmt.Errorf("unable to push: %v", err)
+	}
+	return nil
+}
+
+func getLastBumpCommit(gerritAuthor, commitTag string) (string, error) {
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+
+	if err := Call(&outBuf, &errBuf, gitCmd, "log", fmt.Sprintf("--author=%s", gerritAuthor), fmt.Sprintf("--grep=%s", commitTag), "-1", "--format='%H'"); err != nil {
+		return "", errors.New("error running git command")
+	}
+
+	return outBuf.String(), nil
+}
+
+// getChangeId generates a change ID for the gerrit PR that is deterministic
+// rather than being random as is normally preferable.
+// In particular this chooses a change ID by hashing the last commit by the
+// robot with a given string in the commit message (This string will be added to all autobump commit messages)
+// if there is no commit by the robot with this commit tag, we assume that the job has never run, or that the robot/commit tag has changed
+// in either case, the deterministic ID is generated by just hashing a string of the author + commit tag
+func getChangeId(gerritAuthor, commitTag string) (string, error) {
+	lastBumpCommit, err := getLastBumpCommit(gerritAuthor, commitTag)
+	var id string
+	if err != nil {
+		return "", fmt.Errorf("Error getting change Id: %w", err)
+	}
+	if lastBumpCommit != "" {
+		id = "I" + gitHash(lastBumpCommit)
+	} else {
+		// If it is the first time the autobumper has run a commit will not exist with the tag
+		// create a deterministic tag by hashing the tag itself instead of the last commit.
+		id = "I" + gitHash(gerritAuthor+commitTag)
+	}
+	keepHashing, err := foundCommit(id)
+	//While a commit on the base branch exists with this change ID...
+	for keepHashing && err == nil {
+		// Choose another ID by hashing the current ID.
+		id := "I" + gitHash(id)
+		keepHashing, err = foundCommit(id)
+	}
+	if err != nil {
+		return "", fmt.Errorf("unable to get change id, %s", err)
+	}
+	return id, nil
+}
+
+func foundCommit(id string) (bool, error) {
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+
+	if err := Call(&outBuf, &errBuf, gitCmd, "log"); err != nil {
```

Comment:
The output and error of this command shouldn't change, you can run this before line 1048 for once instead of over and over again in a loop, this also make it possible to replace line 1048 through 1057 with:
```
hashAlreadyExist := true
for hashAlreadyExist {
  id := "I" + gitHash(id)
  keepHashing = strings.Contains(outBuf.String(), id)
}
```

Reply:
Done